File: src/VersionUp/CsprojVersionHandler.cs
namespace VersionUp;

using System;
using System.IO;
using System.Xml.Linq;

/// <summary>
/// Handles parsing and updating version elements in MSBuild project and properties files.
/// </summary>
public class CsprojVersionHandler : IVersionFileHandler
{
    /// <inheritdoc />
    public bool CanHandle(string filePath)
    {
        if (string.IsNullOrEmpty(filePath))
        {
            return false;
        }

        string extension = Path.GetExtension(filePath).ToLowerInvariant();
        string fileName = Path.GetFileName(filePath).ToLowerInvariant();

        bool isProject = extension == ".csproj" || extension == ".fsproj" || extension == ".vbproj";
        bool isBuildProps = fileName == "directory.build.props" || fileName == "directory.build.targets";

        return isProject || isBuildProps;
    }

    /// <inheritdoc />
    public string? GetVersion(string fileContent)
    {
        if (string.IsNullOrWhiteSpace(fileContent))
        {
            return null;
        }

        try
        {
            XDocument doc = XDocument.Parse(fileContent);
            XElement? versionElement = FindVersionElement(doc);

            return versionElement?.Value;
        }
        catch
        {
            return null;
        }
    }

    /// <inheritdoc />
    public string UpdateVersion(string fileContent, string newVersion)
    {
        if (string.IsNullOrWhiteSpace(fileContent))
        {
            fileContent = "<Project><PropertyGroup></PropertyGroup></Project>";
        }

        XDocument doc = XDocument.Parse(fileContent);
        XElement? propertyGroup = doc.Root?.Element("PropertyGroup");

        if (propertyGroup == null)
        {
            propertyGroup = new XElement("PropertyGroup");
            doc.Root?.Add(propertyGroup);
        }

        XElement? versionElement = FindVersionElement(doc);

        if (versionElement == null)
        {
            versionElement = new XElement("Version", newVersion);
            propertyGroup.Add(versionElement);
        }
        else
        {
            versionElement.Value = newVersion;
        }

        string result = doc.ToString();

        return result;
    }

    /// <summary>
    /// Searches the XML document for the first <c>&lt;Version&gt;</c> or
    /// <c>&lt;PackageVersion&gt;</c> element inside any <c>&lt;PropertyGroup&gt;</c>.
    /// </summary>
    /// <param name="doc">The parsed XML document.</param>
    /// <returns>The version element, or <see langword="null"/> if not found.</returns>
    private static XElement? FindVersionElement(XDocument doc)
    {
        if (doc.Root == null)
        {
            return null;
        }

        foreach (XElement propertyGroup in doc.Root.Elements("PropertyGroup"))
        {
            XElement? version = propertyGroup.Element("Version");

            if (version != null)
            {
                return version;
            }

            XElement? packageVersion = propertyGroup.Element("PackageVersion");

            if (packageVersion != null)
            {
                return packageVersion;
            }
        }

        return null;
    }
}

File: vscode-extension/src/handlers/CsprojVersionHandler.ts
import * as path from 'path';
import { IVersionFileHandler } from './IVersionFileHandler';

/**
 * Handles parsing and updating version elements in MSBuild project and properties files.
 */
export class CsprojVersionHandler implements IVersionFileHandler {
    public canHandle(filePath: string): boolean {
        if (!filePath) {
            return false;
        }

        const ext = path.extname(filePath).toLowerCase();
        const filename = path.basename(filePath).toLowerCase();

        const isProject = ext === '.csproj' || ext === '.fsproj' || ext === '.vbproj';
        const isBuildProps = filename === 'directory.build.props' || filename === 'directory.build.targets';

        return isProject || isBuildProps;
    }

    public getVersion(fileContent: string): string | null {
        if (!fileContent || !fileContent.trim()) {
            return null;
        }

        const versionMatch = fileContent.match(/<Version>\s*([^<\s]+)\s*<\/Version>/i);
        if (versionMatch) {
            return versionMatch[1];
        }

        const packageVersionMatch = fileContent.match(/<PackageVersion>\s*([^<\s]+)\s*<\/PackageVersion>/i);
        if (packageVersionMatch) {
            return packageVersionMatch[1];
        }

        return null;
    }

    public updateVersion(fileContent: string, newVersion: string): string {
        if (!fileContent || !fileContent.trim()) {
            return `<Project>\n  <PropertyGroup>\n    <Version>${newVersion}</Version>\n  </PropertyGroup>\n</Project>`;
        }

        if (/<Version>\s*[^<\s]+\s*<\/Version>/i.test(fileContent)) {
            return fileContent.replace(/(<Version>\s*)[^<\s]+(\s*<\/Version>)/i, `$1${newVersion}$2`);
        }

        if (/<PackageVersion>\s*[^<\s]+\s*<\/PackageVersion>/i.test(fileContent)) {
            return fileContent.replace(/(<PackageVersion>\s*)[^<\s]+(\s*<\/PackageVersion>)/i, `$1${newVersion}$2`);
        }

        const propertyGroupMatch = fileContent.match(/<PropertyGroup>/i);
        if (propertyGroupMatch && propertyGroupMatch.index !== undefined) {
            const insertIndex = propertyGroupMatch.index + propertyGroupMatch[0].length;
            return fileContent.slice(0, insertIndex) + `\n    <Version>${newVersion}</Version>` + fileContent.slice(insertIndex);
        }

        const projectMatch = fileContent.match(/<Project[^>]*>/i);
        if (projectMatch && projectMatch.index !== undefined) {
            const insertIndex = projectMatch.index + projectMatch[0].length;
            return fileContent.slice(0, insertIndex) + `\n  <PropertyGroup>\n    <Version>${newVersion}</Version>\n  </PropertyGroup>` + fileContent.slice(insertIndex);
        }

        return fileContent;
    }
}
